ARROW-10665: [Rust] like/nlike utf8 scalar fast paths, bug fixes in like/nlike

Dandandan opened the pull request:
Commonly used patterns '%xxx'  'xxx%' and 'xxx' can use faster methods from Rust standard lib instead.

```
like_utf8 scalar equals time:   [828.13 us 830.08 us 832.39 us]                                    
                        change: [-43.306% -42.962% -42.610%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)

like_utf8 scalar ends with                                                                            
                        time:   [927.93 us 929.31 us 930.88 us]
                        change: [-59.220% -59.149% -59.082%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

like_utf8 scalar starts with                                                                            
                        time:   [930.96 us 931.70 us 932.63 us]
                        change: [-43.537% -43.432% -43.325%] (p = 0.00 < 0.05)
                        Performance has improved.
```

Also tried fast path for contains (`%xxx%`), but that was (almost 2 times) slower than using the regex.

Dandandan commented:
Also looks like there is a bug in the existing implementation (missing anchors), maybe that makes regex faster again. Will follow up later.

jorgecarleitao reviewed (approved):
LGTM. Thanks a lot, @Dandandan .

AFAI can tell, this does not require new tests because it is already covered by the existing tests.

Dandandan commented:
I think it can use some additional tests as I think there is a bug in the current implementation. Will probably have some time tomorrow.

Dandandan commented:
@jorgecarleitao

Added test suites and also fixed existing bugs in like / nlike, also the non scalar ones. Before patterns like "arr", "%arr" "arr%", "arr_", etc were matching too much because of the missing anchors in the regex.

Added benches / fast path to nlike as well.

Dandandan commented:
Possible further optimization for scalar contains ("%string%") would be to look at a different implementation like this crate https://docs.rs/twoway/0.2.1/twoway/ but not sure if it's worth bringing in, regex is already quite fast for contains.
